compiler/AST/AggregateType.cpp

```diff
@@ -853,6 +853,24 @@ AggregateType* AggregateType::getNewInstantiation(Symbol* sym) {
   retval->substitutions.copy(substitutions);
 
   if (field->hasFlag(FLAG_PARAM) == true) {
+    Type* fieldType = NULL;
+    if (field->defPoint->exprType) {
+      fieldType = field->defPoint->exprType->typeInfo();
+    }
+
+    // Sometimes 'sym' might be a different type from the field. For example,
+    // the literal '42' is an int(64), and the field might be a uint(64). In
+    // such cases, we need to coerce to a new symbol that will be placed in
+    // the substitutions map.
+    if (fieldType != NULL  &&
+        fieldType != dtUnknown &&
+        fieldType != sym->getValType()) {
+      Immediate coerce = getDefaultImmediate(fieldType);
+      Immediate* from = toVarSymbol(sym)->immediate;
+      coerce_immediate(from, &coerce);
```

Comment:
What happens if the types can't be coerced?

Reply:
From what I can tell the compiler will fail before this point with a clean-ish error message if the param value and type are incompatible:
```
error: unresolved type specifier 'R(type uint(64), -100)'
candidates are: R(type T, param p: T)
```

Reply:
Okay, sounds good!